Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp

```diff
@@ -277,6 +277,54 @@ void OpenXRCoordinator::scheduleAnimationFrame(WebPageProxy& page, std::optional
         });
 }
 
+void OpenXRCoordinator::waitForWebGLRenderCompletion(WTF::UnixFileDescriptor&& fenceFD)
+{
+    auto& display = m_glContext->display();
+    auto isEGL15 = display.eglCheckVersion(1, 5);
+    EGLSync sync { nullptr };
+    Vector<EGLAttrib> attributes = {
+        EGL_SYNC_NATIVE_FENCE_FD_ANDROID, WTFMove(fenceFD).release(),
+        EGL_NONE
+    };
+    if (isEGL15)
+        sync = eglCreateSync(display.eglDisplay(), EGL_SYNC_NATIVE_FENCE_ANDROID, attributes.span().data());
+    else {
+        Vector<EGLint> intAttributes = attributes.map<Vector<EGLint>>([] (EGLAttrib value) {
+            return value;
+        });
+        sync = eglCreateSyncKHR(display.eglDisplay(), EGL_SYNC_NATIVE_FENCE_ANDROID, intAttributes.span().data());
+    }
+    if (!sync) {
+        RELEASE_LOG(XR, "Failed to create EGL sync");
+        return;
+    }
+
+    static bool supportsServerWait { false };
+#if HAVE(GL_FENCE)
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [&display, &isEGL15] {
+        const auto& extensions = display.eglExtensions();
+        supportsServerWait = isEGL15 ? true : extensions.KHR_wait_sync;
+    });
+#endif
+    if (supportsServerWait) {
+        if (isEGL15)
+            eglWaitSync(display.eglDisplay(), sync, 0);
+        else
+            eglWaitSyncKHR(display.eglDisplay(), sync, 0);
+    } else {
+        if (isEGL15)
+            eglClientWaitSync(display.eglDisplay(), sync, 0, EGL_FOREVER);
+        else
+            eglClientWaitSyncKHR(display.eglDisplay(), sync, 0, EGL_FOREVER_KHR);
```

Comment:
I'm not sure this is a good idea, if server wait is not supported, it's better to wait in the web process or asynchronously poll the fd, because this client wait will block the UI process. For now we can probably assume that if exportable is supported, server wait is too.